proposals/0000-updated-integers.md

```diff
@@ -0,0 +1,687 @@
+
+The prototype implementation, available
+[here](https://github.com/apple/swift/blob/master/test/Prototypes/Integers.swift.gyb)
+contains a `DoubleWidth` generic type that uses two values of any
+`FixedInteger` type to represent a value of twice the width, demonstrating the
+suitability of the proposed model for generic programming.
+
+
+### A note on bit shifts
+
+This proposal introduces the concepts of `smart shifts` and `masking shifts`.
+
+The semantics of shift operations are [often
+undefined](http://llvm.org/docs/LangRef.html#bitwise-binary-operations) in case
+of negative shift or overshift. Smart shifts, implemented by `>>` and `<<`, are
+designed to address this problem and always behave in a well defined way, as
+shown in the examples below:
+
+- `x << -2` is equivalent to `x >> 2`
+
+- `(1 as UInt8) >> 42)` will evaluate to `0`
+
+- `(1 as Int8) >> 42)` will evaluate to `0xff` or `-1`
+
+Masking shifts, implemented by `&>>` and `&<<`, on the other hand, will *not*
+perform any bounds checks and compile to the most efficient code, but are
+subjected to the undefined behavior mentioned above.
+
+
+## Detailed design
+
+### Protocols
+
+#### `Arithmetic`
+
+The `Arithmetic` protocol declares methods backing binary arithmetic operators,
+such as `+`, `-` and `*`; and their mutating counterparts.
+
+It provides a suitable basis for arithmetic on scalars such as integers and
+floating point numbers.
+
+Both mutating and non-mutating operations are declared in the protocol, however
+only the mutating ones are required, as non-mutating are provided by the
+protocol extension.
+
+```Swift
+public protocol Arithmetic : Equatable, IntegerLiteralConvertible {
+  init()
+
+  func adding( rhs: Self) -> Self
+  func subtracting( rhs: Self) -> Self
+  func multiplied(by rhs: Self) -> Self
+  func divided(by rhs: Self) -> Self
+
+  mutating func add( rhs: Self)
+  mutating func subtract( rhs: Self)
+  mutating func multiply(by rhs: Self)
+  mutating func divide(by rhs: Self)
+}
+```
+
+#### `SignedArithmetic`
+
+The `SignedArithmetic` protocol is for numbers that can be negated.
+
+```Swift
+public protocol SignedArithmetic : Arithmetic {
+  func negate() -> Self
+}
+```
+
+#### `Integer`
+
+The `Integer` protocol is the basis for all the integer types provided by the
+standard library.
+
+The `isEqual(to:)` and `isLess(than:)` methods provide implementations for
+`Equatable` and `Comparable` protocol conformances. Similar to how arithmetic
+operations are dispatched in `Arithmetic`, `==` and `<` operators for
+homogeneous comparisons are implemented as generic free functions invoking the
+`isEqual(to:)` and `isLess(than:)` protocol methods respectively.
+
+This protocol adds 3 new initializers to the parameterless one inherited from
+`Arithmetic`. These initializers allow to construct values of one type from
+instances of any other type conforming to `Integer`, using different
+strategies:
+
+  - Assume the value is representatble in `Self`
```

Comment:
on Tue Jun 21 2016, Maxim Moiseev <notifications-AT-github.com> wrote:

>> +
>> +The `Integer` protocol is the basis for all the integer types provided by the
>> +standard library.
>> +
>> +The `isEqual(to:)` and `isLess(than:)` methods provide implementations for
>
>> +`Equatable` and `Comparable` protocol conformances. Similar to how arithmetic
>> +operations are dispatched in `Arithmetic`, `==` and `<` operators for
>> +homogeneous comparisons are implemented as generic free functions invoking the
>> +`isEqual(to:)` and `isLess(than:)` protocol methods respectively.
>> +
>> +This protocol adds 3 new initializers to the parameterless one inherited from
>> +`Arithmetic`. These initializers allow to construct values of one type from
>> +instances of any other type conforming to `Integer`, using different
>> +strategies:
>> +
>> +  - Assume the value is representatble in `Self`
>
> It is the implementation detail. We will trap, 

In that case you have to say it's a precondition on the caller.  The
point is that “assume the value is representable” might just mean
“truncate the bits” or something.

> but should it be described here?

-- 
Dave